test/apps/iodemo/io_demo.cc

```diff
@@ -202,19 +236,65 @@ class IoDemoRandom {
         return size;
     }
 
+    static void setMemoryType(ucs_memory_type_t memory_type)
+    {
+        _memory_type = memory_type;
+    }
+
 private:
     template <typename T>
     static inline void fill(unsigned &seed, T *buffer, size_t count) {
+#ifdef HAVE_CUDA
+        T temp;
+#endif
+
         for (size_t i = 0; i < count; ++i) {
-            buffer[i] = rand<T>(seed);
+            switch (_memory_type) {
+#ifdef HAVE_CUDA
+            case UCS_MEMORY_TYPE_CUDA:
+                temp = rand<T>(seed);
+                cudaMemcpy(&buffer[i], &temp, sizeof(T), cudaMemcpyDefault);
+                break;
+            case UCS_MEMORY_TYPE_CUDA_MANAGED:
+#endif
+            case UCS_MEMORY_TYPE_HOST:
+                buffer[i] = rand<T>(seed);
+                break;
+            default:
+                /* Unreachable - would fail in ctor */
+                abort();
+            }
         }
     }
 
     template <typename T>
     static inline size_t validate(unsigned &seed, const T *buffer, size_t count) {
+#ifdef HAVE_CUDA
+        T expected_value, actual_value;
+#endif
+
         for (size_t i = 0; i < count; ++i) {
-            if (buffer[i] != rand<T>(seed)) {
-                return i;
+            switch (_memory_type) {
+#ifdef HAVE_CUDA
+            case UCS_MEMORY_TYPE_CUDA:
+                expected_value = rand<T>(seed);
+                cudaMemcpy(&actual_value, &buffer[i], sizeof(T),
+                           cudaMemcpyDefault);
```

Comment:
if we calculate sequence on host, maybe do it in separate loop? then we can call cudaMemcpy once for count elements

Reply:
Try:
1. Calculate & validate on CUDA. If not:
2. Calculate on CUDA, copy all to host then validate.